grpc/src/main/java/com/linecorp/armeria/server/grpc/GrpcDecoratingService.java

```diff
@@ -42,43 +50,74 @@ final class GrpcDecoratingService extends SimpleDecoratingHttpService implements
-    private final Map<String, HttpService> pathToDecorators;
-
-    /**
-     * A pair of a fully qualified name of the method (e.g. 'armeria.grpc.sample.SampleService/UnaryCall')
-     * and decorators that are extracted from `@Decorator` and composite already.
-     */
-    private final Map<String, HttpService> methodToDecorators;
-
+    private final HandlerRegistry handlerRegistry;
     private final boolean lookupMethodFromAttribute;
 
-    GrpcDecoratingService(GrpcService delegate, Map<String, HttpService> pathToDecorators,
-                          Map<String, HttpService> methodToDecorators, boolean lookupMethodFromAttribute) {
+    @Nullable
+    private Map<ServerMethodDefinition<?, ?>, HttpService> decorated;
+
+    GrpcDecoratingService(GrpcService delegate, HandlerRegistry handlerRegistry,
+                          boolean lookupMethodFromAttribute) {
         super(delegate);
         this.delegate = delegate;
-        this.pathToDecorators = pathToDecorators;
-        this.methodToDecorators = methodToDecorators;
+        this.handlerRegistry = handlerRegistry;
         this.lookupMethodFromAttribute = lookupMethodFromAttribute;
     }
 
+    @Override
+    public void serviceAdded(ServiceConfig cfg) throws Exception {
+        super.serviceAdded(cfg);
+        final Map<ServerMethodDefinition<?, ?>, List<DecoratorAndOrder>> registryDecorators =
+                handlerRegistry.decorators();
+
+        final Builder<ServerMethodDefinition<?, ?>, HttpService> builder = ImmutableMap.builder();
+
+        for (Entry<ServerMethodDefinition<?, ?>, List<DecoratorAndOrder>> entry
+                : registryDecorators.entrySet()) {
+            final List<? extends Function<? super HttpService, ? extends HttpService>> decorators =
+                    entry.getValue()
+                         .stream()
+                         .map(DecoratorAndOrder::decorator)
+                         .collect(toImmutableList());
+            builder.put(entry.getKey(), applyDecorators(decorators, delegate));
+        }
+        decorated = builder.build();
+    }
+
+    private static HttpService applyDecorators(
+            List<? extends Function<? super HttpService, ? extends HttpService>> decorators,
+            HttpService delegate) {
+        Function<? super HttpService, ? extends HttpService> decorator = Function.identity();
+        for (int i = decorators.size() - 1; i >= 0; i--) {
+            decorator = decorator.andThen(decorators.get(i));
+        }
+        return decorator.apply(delegate);
+    }
+
     @Override
     public HttpResponse serve(ServiceRequestContext ctx, HttpRequest req) throws Exception {
-        HttpService decoratedService = lookup(ctx);
-        if (decoratedService != null) {
-            return decoratedService.serve(ctx, req);
+        assert decorated != null;
+        ServerMethodDefinition<?, ?> methodDefinition =
+                lookupMethodFromAttribute ? ctx.attr(RESOLVED_GRPC_METHOD) : null;
```

Comment:
Would it be weird to move `RESOLVED_GRPC_METHOD` to `GrpcService.method(ctx)`?

Reply:
I found out that we only need to set the attribute when `HttpJsonTranscodingService` is enabled.
So I changed to use the attribute in `FramedGrpcService` and `GrpcDecoratingService` just delegates the method call to `FramedGrpcService`. PTAL. 😄